File: DP/206-reverse-linked-list/reverse-linked-list.java
class Solution {
    public ListNode reverseList(ListNode head) {
        return reverse(head);
    }
    private ListNode reverse(ListNode head){
        ListNode curr = head, prev = null;
        while(curr != null){
           ListNode nextNode = curr.next;
           curr.next = prev;
           prev = curr;
           curr = nextNode;
        }
        return prev;
    }
}

File: DP/206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* reverse(ListNode* curr, ListNode* pre){
        if(curr==NULL){
            return pre;
        }
        ListNode* temp = curr->next;
        curr->next = pre;
        pre = curr;
        curr = temp;
        return reverse(curr, pre);
    }
    ListNode* reverseList(ListNode* head) {
        ListNode* curr = head;
        ListNode* pre = NULL;

      return  reverse(curr, pre);
          
    } 
};